### PR TITLE
test_return_object_rarelabelencoder

### PR DESCRIPTION
**Pull request by KarthikKothareddy**

### CODE REVIEW

File: feature_engine/feature_selection.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Authors: Soledad Galli <solegalli@protonmail.com>
+# License: BSD 3 clause
```
Comment (solegalli) [inline feature_engine/feature_selection.py:2]: please erase rows 1 and 2

File: feature_engine/feature_selection.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import numpy as np
+import pandas as pd
+import warnings
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from feature_engine.dataframe_checks import _is_dataframe
+from feature_engine.variable_manipulation import _define_variables
+
+
+class FeatureEliminator(BaseEstimator, TransformerMixin):
```
Comment (solegalli) [inline feature_engine/feature_selection.py:13]: is the name FeatureEliminator clear enough? 
I think I would prefer to call the transformer **DropFeatures**? 

Eliminator gives me the sensation that there is some sort of fancy algorithm running, and we are only removing the features that the user inputs. What do you think?

File: feature_engine/feature_selection.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        ----------
+
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The input dataframe on which the feature elimination has to be performed
+        y: None
+            y is not needed for this transformer
+
+        """
+        # check input dataframe
+        X = _is_dataframe(X)
```
Comment (solegalli) [inline feature_engine/feature_selection.py:48]: the fit method needs to return at least 1 attribute with a trailing underscore, for example self.input_shape_, because the trailing underscore is that the is_fitted method checks when executing the transform method, to check that the user used fit to fit the transformer before using transform. I will point to example code as I go on with the review.

Comment (solegalli) [reply feature_engine/feature_selection.py:48]: fit method should include the `self.input_check`, as here:
https://github.com/solegalli/feature_engine/blob/master/feature_engine/variable_transformers.py#L222

I think we should also include a check to corroborate that the variables entered by the user are present in the dataframe. If not, return an error. 

What I am not sure is if we should also check absence of na in the training set. Thoughts?

File: feature_engine/feature_selection.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    features_to_drop : str or list, default=None
+        Desired variable/s to be dropped from the dataframe
+
+    """
+
+    def __init__(self, features_to_drop=None):
+
+        if not any(isinstance(features_to_drop, t) for t in [str, list]):
+            raise ValueError("features_to_drop must be a string or list object")
+
```
Comment (solegalli) [inline feature_engine/feature_selection.py:31]: `_define_variables` checks if the input is a string or a list, so I don't think we need to include these 2 lines of code.

File: feature_engine/feature_selection.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    def fit(self, X, y=None):
+        """
+        Verifies that the passed input X if of the type pandas dataframe
+
+        Parameters
+        ----------
+
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The input dataframe on which the feature elimination has to be performed
+        y: None
```
Comment (solegalli) [inline feature_engine/feature_selection.py:43]: please add an empty line between parameters, otherwise, it does not look nice when the docs are built.

File: feature_engine/feature_selection.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        X: pandas dataframe
+            The input dataframe on which the feature elimination has to be performed
+
+        Returns
+        -------
+        X_transformed: pandas dataframe of shape = [n_samples, n_features - len(features_to_drop)]
+            The transformed dataframe with subset of variables.
+
+        """
+        return X.drop(columns=self.features).copy()
```
Comment (solegalli) [inline feature_engine/feature_selection.py:67]: in the transform method we need to check that the transformer is fitted, that the X provided by the user is a dataframe, and that the dataframe matches the size of that provided to train the transformer. This is to help the user keep its pipelines tidy. See here for an example:
https://github.com/solegalli/feature_engine/blob/master/feature_engine/base_transformers.py#L155

In the return, please return just the final output and not an operation. Doesn't look too nice in my opinion.

File: tests/test_categorical_encoder.py [hunk 7]
```diff
@@ -446,7 +446,7 @@ def test_OneHotCategoricalEncoder(dataframe_enc_big, dataframe_enc_big_na):
         encoder.transform(dataframe_enc_big_na)
 
 
-def test_RareLabelEncoder(dataframe_enc_big, dataframe_enc_big_na):
+def test_RareLabelEncoder(dataframe_enc_big, dataframe_enc_big_na, dataframe_enc_rare):
     # test case 1: defo params, automatically select variables
```
Comment (solegalli) [inline tests/test_categorical_encoder.py:450]: could you make this file a separate PR?

File: tests/test_feature_selection.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
+# Authors: Soledad Galli <solegalli1@gmail.com>
+# License: BSD 3 clause
+import pytest
+import pandas as pd
+
+from feature_engine.feature_selection import FeatureEliminator
+
```
Comment (solegalli) [inline tests/test_feature_selection.py:7]: could you change the format of the tests so that each test case is actually a separate test? see for example this file:
https://github.com/solegalli/feature_engine/blob/master/tests/test_wrappers.py

I will be reformatting all the tests to this, because it is much simple to know which test is falling if they are individual. Each test should have a name describing that it is testing.

File: tests/test_feature_selection.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    with pytest.raises(ValueError):
+        transformer = FeatureEliminator(features_to_drop=99)
+        X = transformer.fit_transform(dataframe_vartypes)
+
+    # test case 5: passing a different input than dataframe
+    with pytest.raises(TypeError):
+        transformer = FeatureEliminator(features_to_drop=["Name"])
+        X = transformer.fit_transform({"Name": ["Karthik"]})
+
+    # test case 6: dropping all columns
```
Comment (solegalli) [inline tests/test_feature_selection.py:48]: this is actually a good point. Should we not warn the user that they are dropping all the variables and returning an empty dataframe in the FeatureEliminator?

File: tests/test_feature_selection.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    transformer = FeatureEliminator(features_to_drop=list(dataframe_vartypes.columns))
+    X = transformer.fit_transform(dataframe_vartypes)
+
+    assert len(X) == len(dataframe_vartypes)
+    assert X.shape == (4, 0)
+
+    # test case 7: passing an empty list
+    transformer = FeatureEliminator(features_to_drop=[])
+    X = transformer.fit_transform(dataframe_vartypes)
```
Comment (solegalli) [inline tests/test_feature_selection.py:57]: I think we should catch this at the transformer. A check that the list is not empty, if not, return an error. Could you please add it?

**Comment by solegalli** [comment]
hi @KarthikKothareddy this is great work! Thank you so much for the contribution. I have a few suggestions here and there, and a few checks that need to be added in to the transformer. Could you please have a look? Many thanks!

**Comment by solegalli** [comment]
@renamed to PR to something more descriptive of the issue.

**Comment by solegalli** [comment]
@KarthikKothareddy  I just noticed that there is another PR for the feature eliminator, and this pr should be only for the rarelabel encoder. I suggest you follow the steps I indicate in PR96. Then we close this PR. And you make a new PR just with the change for the rare label encoder test. Does that sound doable? 

Please make sure you check out each feature branch from master. and that master is in sync with the remote master in feature-engine's repo.

Thank you!

File: tests/test_feature_selection.py (hunk 8 above)
Comment (KarthikKothareddy) [reply tests/test_feature_selection.py:7]: Absolutely, I kind of thought along the same lines but didn't wanted to mess with the existing structure of tests. I will reformat them.

File: tests/test_categorical_encoder.py (hunk 7 above)
Comment (KarthikKothareddy) [reply tests/test_categorical_encoder.py:450]: I was under the assumption that this will be treated as a separate one. I will drop it and create another.

File: feature_engine/feature_selection.py (hunk 2 above)
Comment (KarthikKothareddy) [reply feature_engine/feature_selection.py:13]: I am very bad in naming things so, I think DropFeatures makes sense. If you're okay with DropFeatures I will go ahead and rename it.

File: feature_engine/feature_selection.py (hunk 3 above)
Comment (KarthikKothareddy) [reply feature_engine/feature_selection.py:48]: > the fit method needs to return at least 1 attribute with a trailing underscore, for example self.input_shape_, because the trailing underscore is that the is_fitted method checks when executing the transform method, to check that the user used fit to fit the transformer before using transform. I will point to example code as I go on with the review.

I will make sure that is added.

Comment (KarthikKothareddy) [reply feature_engine/feature_selection.py:48]: > fit method should include the `self.input_check`, as here:
> https://github.com/solegalli/feature_engine/blob/master/feature_engine/variable_transformers.py#L222
> 
> I think we should also include a check to corroborate that the variables entered by the user are present in the dataframe. If not, return an error.
> 
> What I am not sure is if we should also check absence of na in the training set. Thoughts?

As far having a check for variables that doesn't exist, pandas drop() does raise a "KeyError" if it finds a variable not present in the dataframe. I also have a test case to check this (test case 2). Will that be sufficient or do you need an explicit check?
